File: migration/pset6/old/credit/credit.py
# credit.py
# This program checks if a credit
# card number entered by a user is
# a valid credit card number by
# using Luhn's Algorithm to verify
# the checksum, and it identifies if
# it's an American Express, Mastercard,
# Visa credit card, or if the card is
# invalid.
from cs50 import get_int

# prompts the user for a card number
cc_num = get_int("Card Number: ")

# checks if the card number is valid or not
cc_cpy = cc_num
i = 0
AmX = False
Vis = False
Mas = False

# calculates number of digits entered
while cc_cpy:
    cc_cpy //= 10
    i += 1

if i == 14 or (i < 13 or i > 16):
    print("INVALID")

elif i == 13:
    p = i - 1
    if cc_num // 10**p != 4:
        print("INVALID")
    elif cc_num // 10**p == 4:
        Vis = True

elif i == 16:
    p = i - 1
    q = i - 2
    if (cc_num // 10**p != 4) and cc_num // 10**p != 5:
        print("INVALID")
    elif (cc_num // 10**p == 4):
        Vis = True
    elif cc_num // 10**p == 5:
        if (cc_num // 10**q < 51) or (cc_num // 10**q > 55):
            print("INVALID")
        else:
            Mas = True

elif i == 15:
    p = i - 2
    if (cc_num // 10**p != 34) and (cc_num // 10**p != 37):
        print("INVALID")
    else:
        AmX = True

# checks validity using Luhn's Algorithm
i, odd, even = 0, 0, 0
two, ten, nine = 2, 10, 9
cc_cpy = cc_num
while cc_cpy:
    if i % two == 0:
        odd += cc_cpy % ten
    elif i % two != 0:
        x = (cc_cpy % ten) * two
        if x > nine:
            even += (x // ten) + (x % ten)
        else:
            even += x
    # updates number
    cc_cpy //= ten
    # increments
    i += 1

total = odd + even

if total % ten == 0:
    if AmX:
        print("AMEX")
    elif Mas:
        print("MASTERCARD")
    elif Vis:
        print("VISA")
else:
    print("INVALID")

File: CS50/pset1/old/credit.c
/*
 * credit.c
 *
 * This program uses Luhn's Algorithm
 * to check if the credit card number
 * entered by a user is valid or not.
 *
 * It also identifies if the card is
 * a VisaCard, Mastercard, or American
 * Express card. The program identifies
 * valid cards outside of these three as
 * "Unknown"
 */
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    /*
     * Prompts the user for a credit
     * card number.
     */
    long long cc_number;

    do
    {
        cc_number = get_long_long("Number: ");
    }
    // Loops while digits are < 10 or > 16
    while ((cc_number < 1000000000) || (cc_number > 9999999999999999));

    /*
     * Determines the length of the card number.
     * copy_ccnmbr is a copy of the credit card
     * number used to determine the length so that
     * the original number is not modified
     */
    long long length = 0, copy_ccnmbr = cc_number;

    while (copy_ccnmbr != 0)
    {
        copy_ccnmbr /= 10;
        length++;
    }

    // Calculates the checksum of the card
    long long i = 0, j = 0, x = 0, y = 0, total = 0;

    for (i = 1, j = 1; i <= length; i++, j = j * 10)
    {
        if ((i % 2) == 0) // Identifies even-numbered positions
        {
            x = ((cc_number / j) % 10);
            y = x * 2;

            if (y > 9)
            {
                total = total + (y / 10) + (y % 10); // Sums up digits of a 2-digit number
                x = 0;
            }
            else
            {
                total = total + y;
                x = 0;
            }
        }
        else
        {
            x = ((cc_number / j) % 10);
            total = total + x;
            x = 0;
        }
    }

    // Determines the company identifier
    long long coy_ID = cc_number;

    if ((total % 10) == 0)
    {
        while (coy_ID >= 100)
        {
            coy_ID = coy_ID / 10;
        }

        if (((length == 13) || (length == 16)) && ((coy_ID > 39) && (coy_ID < 50)))
        {
            printf("Visa\n");
        }

        else if ((length == 16) && ((coy_ID > 50) && (coy_ID < 56)))
        {
            printf("Mastercard\n");
        }

        else if ((length == 15) && ((coy_ID == 34) || (coy_ID == 37)))
        {
            printf("American Express\n");
        }

        else
        {
            printf("Unknown\n");
        }
    }

    else
    {
        printf("Invalid\n");
    }
}